services/src/kilda-pce/src/main/java/org/openkilda/pce/finder/BestWeightAndShortestPathFinder.java

```diff
@@ -286,6 +305,69 @@ private void restoreEdge(Edge edge) {
         return (bestPath != null) ? bestPath.parentPath : new LinkedList<>();
     }
 
+    /**
+     * Finds a path whose weight is less than maxWeight and as close to maxWeight as possible.
+     *
+     * @return An ordered list that represents the path from start to end, or an empty list
+     */
+    private List<Edge> getPath(Node start, Node end, WeightFunction weightFunction, long maxWeight) {
+        long bestWeight = Long.MAX_VALUE;
+        SearchNode bestPath = null;
+
+        Deque<SearchNode> toVisit = new LinkedList<>();
+        Map<Node, SearchNode> visited = new HashMap<>();
+
+        toVisit.add(new SearchNode(weightFunction, start, allowedDepth, 0, emptyList()));
+
+        while (!toVisit.isEmpty()) {
+            SearchNode current = toVisit.pop();
+
+            // Leave if the path contains this node
+            if (current.containsSwitch(current.dstSw.getSwitchId())) {
+                continue;
+            }
+
+            // Shift the current weight relative to maxWeight
+            long shiftedCurrentWeight = Math.abs(maxWeight - current.parentWeight);
+
+            // Determine if this node is the destination node.
+            if (current.dstSw.equals(end)) {
+                // We found the destination
+                if (shiftedCurrentWeight < bestWeight && current.parentWeight < maxWeight) {
+                    // We found a best path. If we don't get here, then the entire graph will be
+                    // searched until we run out of nodes or the depth is reached.
+                    bestWeight = shiftedCurrentWeight;
+                    bestPath = current;
+                }
+                // We found dest, no need to keep processing
+                continue;
+            }
+
+            // Stop processing entirely if we've gone too far, or over maxWeight
+            if (current.allowedDepth <= 0 || current.parentWeight > maxWeight) {
+                continue;
+            }
+
+            // Otherwise, if we've been here before, see if this path is better
+            SearchNode prior = visited.get(current.dstSw);
+            if (prior != null && shiftedCurrentWeight >= Math.abs(maxWeight - prior.parentWeight)) {
+                continue;
+            }
+
+            // Either this is the first time, or this one has less weight .. either way, this node should
+            // be the one in the visited list
+            visited.put(current.dstSw, current);
+
+            // At this stage .. haven't found END, haven't gone too deep, and we are not over weight.
+            // So, add the outbound isls.
+            current.dstSw.getOutgoingLinks().stream()
+                    .sorted(Comparator.comparing(edge -> edge.getDestSwitch().getSwitchId()))
+                    .forEach(edge -> toVisit.add(current.addNode(edge)));
+        }
+
+        return (bestPath != null) ? bestPath.parentPath : new LinkedList<>();
+    }
```

Comment:
Algorithm with max weight is very similar to previous one. Maybe we can generalize it using different weight shift and path weight comparison functions to avoid code duplication?

Reply:
Algorithm with max weight looks similar, but we are searching path in another way.